api/v1beta1/clusterclass_types.go

```diff
@@ -637,6 +649,11 @@ type LocalObjectTemplate struct {
 	Ref *corev1.ObjectReference `json:"ref"`
 }
 
+// NamingStrategy defines a naming strategy for templated objects.
+type NamingStrategy struct {
```

Comment:
Hey very good ask!

I just had a short discussion with @sbueringer about this and I think there are three options:


Option 1 (as currently is on this PR):

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

Option 2: similar to what CAPI already does for variables

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```


Option 3 (what you proposed): similar to Option 2 but adding a type:

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name:
        type: Template
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName:
        type: Template
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

I lean towards Option 2. It is also flexible to move to option 3 in future if we introduce other variants..

Note: This also adds a `machineInfrastructureTemplateName`, to show how it could get expanded to other resources. In scope for implementation for now is only `name`.

Reply:
+1 for Option 2

Reply:
Option 2 seems the cleanest to me as well - thanks for working on the examples!

Reply:
I prefer 2 as well, but what about 

```
spec:
  ...
  controlPlane:
    namingStrategy:
      - field: metadata.name
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      - field:  spec.machineTemplate.infrastructureRef.Name
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"`
```

so we are a little bit more explicit about where the generated names will pop up in the controlPlane

Reply:
This looks more like a generic engine to patch arbitrary fields to me to be honest

Reply:
From discussions: there's a 4th option which would be more consistent compared to the current structure of the ClusterClass object:

```
spec:
  ...
  controlPlane:
    namingStrategy:
      template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplate:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

Reply:
Sounds good to me

Reply:
+1 from me